docs/versionhistory.rst

```diff
@@ -7,6 +7,13 @@ This library adheres to `Semantic Versioning 2.0 <http://semver.org/>`_.
 
 - **BACKWARDS INCOMPATIBLE** Replaced AnyIO's own ``ExceptionGroup`` class with the PEP
   654 ``BaseExceptionGroup`` and ``ExceptionGroup``
+- **BACKWARDS INCOMPATIBLE** Changes to cancellation semantics:
+
+  - Any exceptions raising out of a task groups are now nested inside an
+    ``ExceptionGroup`` (or ``BaseExceptionGroup`` if one or more ``BaseException`` were
+    included)
+  - Fixed task group not raising a cancellation exception on asyncio at exit if no child
+    tasks were spawned and an outer cancellation scope had been cancelled before
```

Comment:
Trio would say that a Cancelled exception means some work was interrupted. In this case, it sounds like no work was interrupted, so I'm not sure it's actually important to raise the exception. Trio does raise the exception, because of its rule that every async operation that can block (including `NurseryManager.__aexit__`) is always both a schedule point and a cancel point, but this has caused problems occasionally and there has been recent discussion of changing it; IIRC consensus was that it would be fine to change but no one has done it yet. Discussion in python-trio/trio#1457.

Reply:
The issue that this refers to was that an AnyIO `TaskGroup` would swallow a `CancelledError` that was meant to cancel the entire task.

Reply:
Edge-triggered cancellation rears its ugly head again!

My intuition is that for fully consistent behavior here you need to treat each asyncio task as implicitly enclosed in a CancelScope that would be cancelled by `task.cancel()`, and include that implicit cancel scope in your logic of whether each cancel scope should swallow a cancellation or forward it outward. You can indirectly detect whether a task-level cancellation has occurred, at least on 3.11+, by counting anyio-originated calls to `task.cancel()` that are not yet uncancelled, and comparing that against `task.cancelling()`.

Making task group exit a cancel point is defensible though. Even if Trio stops doing that, you can easily add it back by putting in an explicit `checkpoint_if_cancelled`.

Reply:
> and include that implicit cancel scope in your logic of whether each cancel scope should swallow a cancellation or forward it outward.

This is actually something I was going to propose for asyncio! I'll be sending that  post to Async-SIG shortly. Meanwhile, I can again look for a workaround for Python < 3.11, but that should not block this PR.